pyo3-derive-backend/src/pymethod.rs

```diff
@@ -6,97 +6,29 @@ use quote::quote;
-             __new__, put it on the struct definition instead",
-        )?,
-        FnType::FnCall => parse_erroneous_text_signature(
-            Some("__call__"),
-            "text_signature not allowed on __call__",
-        )?,
-        FnType::Getter(getter_name) => parse_erroneous_text_signature(
-            getter_name.as_ref().map(|v| &**v),
-            "text_signature not allowed on getter",
-        )?,
-        FnType::Setter(setter_name) => parse_erroneous_text_signature(
-            setter_name.as_ref().map(|v| &**v),
-            "text_signature not allowed on setter",
-        )?,
-    };
-    let doc = utils::get_doc(&meth_attrs, text_signature, true)?;
+    let spec = FnSpec::parse(sig, &mut *meth_attrs, true)?;
 
     Ok(match spec.tp {
-        FnType::Fn => impl_py_method_def(name, doc, &spec, &impl_wrap(cls, name, &spec, true)),
-        FnType::PySelf(ref self_ty) => impl_py_method_def(
-            name,
-            doc,
-            &spec,
-            &impl_wrap_pyslf(cls, name, &spec, self_ty, true),
-        ),
-        FnType::FnNew => impl_py_method_def_new(name, doc, &impl_wrap_new(cls, name, &spec)),
-        FnType::FnCall => impl_py_method_def_call(name, doc, &impl_wrap(cls, name, &spec, false)),
-        FnType::FnClass => impl_py_method_def_class(name, doc, &impl_wrap_class(cls, name, &spec)),
-        FnType::FnStatic => {
-            impl_py_method_def_static(name, doc, &impl_wrap_static(cls, name, &spec))
-        }
-        FnType::Getter(ref getter) => {
-            let takes_py = match &*spec.args {
-                [] => false,
-                [arg] if utils::if_type_is_python(arg.ty) => true,
-                _ => {
-                    return Err(syn::Error::new_spanned(
-                        spec.args[0].ty,
-                        "Getter function can only have one argument of type pyo3::Python!",
-                    ));
-                }
-            };
-            impl_py_getter_def(name, doc, getter, &impl_wrap_getter(cls, name, takes_py))
-        }
-        FnType::Setter(ref setter) => {
-            impl_py_setter_def(name, doc, setter, &impl_wrap_setter(cls, name, &spec))
+        FnType::Fn => impl_py_method_def(&spec, &impl_wrap(cls, &spec, true)),
+        FnType::PySelf(ref self_ty) => {
+            impl_py_method_def(&spec, &impl_wrap_pyslf(cls, &spec, self_ty, true))
         }
+        FnType::FnNew => impl_py_method_def_new(&spec, &impl_wrap_new(cls, &spec)),
+        FnType::FnCall => impl_py_method_def_call(&spec, &impl_wrap(cls, &spec, false)),
+        FnType::FnClass => impl_py_method_def_class(&spec, &impl_wrap_class(cls, &spec)),
+        FnType::FnStatic => impl_py_method_def_static(&spec, &impl_wrap_static(cls, &spec)),
+        FnType::Getter => impl_py_getter_def(&spec, &impl_wrap_getter(cls, &spec)?),
+        FnType::Setter => impl_py_setter_def(&spec, &impl_wrap_setter(cls, &spec)?),
```

Comment:
Now that these function signatures are so similar (and their implementations very similar too) I wonder whether with a bit of refactoring this can be simplified. Perhaps in a separate PR.